### ISSUE TITLE
Occasional reconnect problems

### ISSUE THREAD

**Issue by Dev-Eloper** [vtortola/WebSocketListener#14]
Hi,

I'm experiencing occasional problems when the client has closed the connection (perhaps ungracefully) and then tries to reconnect. The issue is quite difficult to reproduce. I see it about once a week. When it occurs, the symptoms are as follows:

- The client reports that the connection was established successfully
- The server however does not return from WebSocketListener.AcceptWebSocketAsync()
- The server does arrive at WebSocketListener.NegotiateWebSocket(), and in fact this function returns a positive result: it says the socket is connected and no error occurred.
- When I try to connect from a different client, AcceptWebSocketAsync() does return and a connection with the second client is established succesfully.
- As soon as the second client has connected, a new connection attempt by the first client is also accepted.

So, it's quite a mystery to me what is going on here. To me this looks like a bug in the TransformBlock class, because I would expect TransformBlock.ReceiveAsync() to return as soon as NegotiateWebSocket has produced a valid WebSocketNegotiationResult. Perhaps you have some ideas?

**Comment by vtortola** [comment]
Hi,

So basically you say that you see a connection being correctly negotiated that never is yield, however further connections are negotiated and yield correctly, right?

Is this happening when many customers try to connect at the same time? or after a long idle period? or it is completely random? What number of connections per week or day are you having at the moment? Is it a particular browser or client?

When you are calling `AcceptWebSocketAsync` are you doing it using `await` (rather than `.Result`) and in a `try/catch` block ? Are you completely sure that `NegotiateWebSocket` is negotiating well and `result.Error` and `handshake.Error` are `null` at the end of the method? Please note that if there is an exception registered in `handshake.Error` it is not thrown in that method.

Cheers.

**Comment by Dev-Eloper** [comment]
The issue happens when only a single client is connected, then disconnects, then reconnects within a few seconds (I'm not yet sure what the influence is of the interval between the disconnect and the reconnect). I connect a few hundred times a week.
AcceptWebSocketAsync is called with await and in a try/catch block, and with configureAwait(false). Every exception is logged, but in this case I never see any exception.
I have not looked at handshake.Error. Next time I see this issue, I'll make sure to have a look at it. Thanks!

**Comment by Dev-Eloper** [comment]
And concerning your first question: that is correct, insofar that the first connection would not be established until a second connection was established.

**Comment by vtortola** [comment]
I will try to identify what may cause something like that. It is very strange, it is like a connection is lost inside the TransformBlock, and that is very unlikely.

**Comment by Dev-Eloper** [comment]
Thanks. Indeed it looks very strange. Even if handshake.Error would be set, I would still expect TransformBlock.ReceiveAsync() to return. I'll write a test tomorrow that disconnects and connects in quick succession. Hopefully I can reproduce the issue this way.

**Comment by Dev-Eloper** [comment]
I suspect that the scenario must be as follows: 
- Client A begins the negotiation with the server
- Somewhere in the middle of the negotiation, the client is killed
- The server is now stuck in the middle of a negotiation.
- Client A is restarted and starts a new negotiation. This negotiation is successful and the NegotiationResult is appended to the TransformBlock. However, the TransformBlock processes all negotiations in order, so it cannot return from ReceiveAsync until the first negotiation is completed. This should happen when the negotiationTimeout expires. 

I do seem to recall having seen a negotiation timeout at some point, but i couldn't find this in my logs anymore so I thought I was mistaken. I have set the negotiationTimeout to 30 seconds, so in theory a new connection should be possible after this time. I'll try to reproduce the above scenario in a test and see if the negotiationTimeout does indeed fire at the expected moment, and if this gets the server unstuck.

**Comment by Dev-Eloper** [comment]
I'm a bit further. I've adjusted the client a bit so it never completes the handshake. The server then hangs indefinitely in HandshakeAsync(). Although NegotiateWebSocket() tries to wake up the thread when the NegotiationTimeout expires by way of Task.WhenAny(timeoutTask, handshakeTask), it never gets to this point because HandshakeAsync() blocks.

**Comment by Dev-Eloper** [comment]
I was able to solve the problem with HandshakeAsync by making ReadHttpRequest asynchronous:

<code>
private async Task ReadHttpRequestAsync(Stream clientStream, WebSocketHandshake handshake)
        {
            using (var sr = new StreamReader(clientStream, Encoding.ASCII, false, 1024, true))
            {
                String line = await sr.ReadLineAsync().ConfigureAwait(false);

                ParseGET(line, handshake);

                while (!String.IsNullOrWhiteSpace(line = await sr.ReadLineAsync().ConfigureAwait(false)))
                    ParseHeader(line, handshake);
            }
        }
</code>

With this change, NegotiateWebSocket() does arrive at Task.WhenAny(timeoutTask, handshakeTask), causing AcceptWebSocketAsync() to return when the NegotiationTimeout expires.

**Comment by Dev-Eloper** [comment]
I think that in theory, a similar problem could occur in the exception handler of HandShakeAsync(), because this contains another synchronous call, to WriteHttpResponse(), which I think should be WriteHttpResponseAsync(), right?

**Comment by vtortola** [comment]
I have to investigate this. I dont really understand why the negotiation timeout is not working, it should kick in, picked by other thread, throw an exception and close the socket, releasing the thread blocked reading or writting. I will give it a closer look this week. Thanks!

**Comment by Dev-Eloper** [comment]
Well, apparently HandShakeAsync, or any async method for that matter, doesn't actually run asynchronous until the moment it awaits another async method. This is indeed a bit counter-intuitive, but apparently this is the way the async methods have been designed. See "gotcha 1" in this article: http://tomasp.net/blog/csharp-async-gotchas.aspx/

**Comment by vtortola** [comment]
I knew that, but I was expecting the negotiations to run in parallel in the `TransformBlock`. I will run some experiments to check it.

**Comment by Dev-Eloper** [comment]
Indeed, that's the thing that's also still puzzling me. I still don't understand why one blocking connect attempt would cause the subsequent connect attempts to fail as well. Is it possible that the TransformBlock considers the two sockets identical?
Also, why would connecting from a different client suddenly unblock the initial connection attempt? I'm trying to reproduce this  behaviour now, but so far without luck.

BTW, I'm using the following settings:
	PingTimeout = TimeSpan.FromSeconds(10),
	NegotiationTimeout = TimeSpan.FromSeconds(5),
	WebSocketSendTimeout = TimeSpan.FromSeconds(10),
	WebSocketReceiveTimeout = TimeSpan.FromSeconds(1000000),
	ParallelNegotiations = 16,
	NegotiationQueueCapacity = 256,
	TcpBacklog = 64,

In the default settings, the ReceiveTimeout is identical to the NegotiationTimeout, as a result of which you may see different behaviour than what I reported.

**Comment by Dev-Eloper** [comment]
Correction: in my initial settings, when the bug occurred, the NegotationTimeout was 30 seconds.

**Comment by vtortola** [comment]
You were in the right track.

TransformBlock is doing negotiations in parallel but ensuring ordering. The timeout mechanism is not working when the thread gets block, and because the TB is ensuring ordering, cannot yield next item until that one is released.

Your solution is good, since it will place an `await` that will potentially push the rest of the method to another thread. But I think is going to be more elegant to put an [`await Task.Yield()`](http://stackoverflow.com/a/22645114/307976) on the first line of `HandshakeAsync`, and refactor that part a little bit regarding async methods (they should probably be all sync, since the data is already in the `NetworkStream` when reading, and when writing, a `NetworkStream` is a `BufferedStream` so probably is no point to `await` anything). Let me investigate it further.

Thanks a million for your debugging time :)

**Comment by Dev-Eloper** [comment]
Really? I had already dismissed the idea that TransformBlock preserved ordering because it doesn't explain why connecting from a different client would unblock it. For the time being I won't change anything to the code yet and hope the issue reappears, because I would really like to understand what happens when the second client connects.
Regardless whether this really fixes the issue though, it does seem like a good idea to change HandShakeAsync(). I agree that putting await Task.Yield() on the first line is the more elegant solution.

**Comment by Dev-Eloper** [comment]
And you're welcome ;) I really like the library you wrote, so it deserves some extra effort.

**Comment by vtortola** [comment]
I neither find how this causes the first scenario you explained, where an item is just lost in the TB. I mean, what we have found out so far, is that if a connection gets stuck in the middle of the negotation, other netogiations will happen in parallel, but won't be consumed since TB has to wait for the stuck one to finish. But you said that further connection succeed properly, so something else is going on here.
The proper action would be to try to reproduce the problem before touching anything. I will try to create a test client that creates many connections and disconnections and see what happens.

**Comment by Dev-Eloper** [comment]
I reproduced the scenario we do understand just now: I ran a test where a single client was started and then killed after a random interval. At some point, after about 4000 connection attempts, this resulted in an invalid handshake. The NegotiationQueue's OutputCount then started to grow upto 255 (the NegotiationQueueCapacity). After that, no more connection attempts were accepted. Connecting from a different client didn't help at this point, the NegotiationQueue remained blocked.

**Comment by Dev-Eloper** [comment]
Hang on. This has a totally different cause. I appear to have deadlock in my own code.

**Comment by Dev-Eloper** [comment]
OK, fixed the deadlock and reran the test. The issue occurred within a few minutes now. It turns out there is actually a second client involved from the start. I hadn't realized this because it ran on a phone that was inactive. It looks like the second client refuses to complete the handshake, until the moment the phone is activated (I'm not sure why yet). This in combination with the problem in HandshakeAsync and the fact that the TransformBlock preserves the order causes the first client to be unable to connect until the phone is activated on which the second client runs.
So adjusting HandshakeAsync should indeed fix the issue. There is still some weird behaviour in my client app, but you shouldn't be concerned with that.

**Comment by vtortola** [comment]
I will fix the timeout, but still if there are many connections not finishing the handshake, they could slow down considerably the connection rate because the TB is trying to preserver order. Since there is no point in preserving ordering, I will remove the TransformBlock and create one block similar that does not preserve it.

**Comment by Dev-Eloper** [comment]
Good point. I think I fixed that issue in the client, but I'll doublecheck it. Note that even if a block was involved that wouldn't preserve the order, it would probably still have a limited capacity, so lingering handshakes could still prevent other clients from connecting.

**Comment by vtortola** [comment]
Hi,

I did a commit that should fix the issue: https://github.com/vtortola/WebSocketListener/commit/416ba11e7627645e8170880249870eaa648b3696

I removed the `TransformBlock<>`, so now negotiations do not care about order. I tested it as usual and it is working fine, but I want to do another test about heavy connection/disconnection before releasing in Nuget.

Let me know if you have a chance to test it.
